File: 285. Inorder Successor in BST/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderSuccessor(self, root: 'TreeNode', p: 'TreeNode') -> 'TreeNode':
        # reach the node
        parent = None
        node = root
        while node and node.val != p.val:
            if node.val > p.val:
                parent = node
                node = node.left
            else:
                node = node.right
                
        # get the successor
        if node.right:
            node = node.right
            while node.left:
                node = node.left
                
            return node
        else:
            return parent

File: 285. Inorder Successor in BST/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        // find the node
        TreeNode prev = null;
        while (root != null && root.val != p.val) {
            if (root.val > p.val) {
                prev = root;
                root = root.left;
            } else {
                root = root.right;
            }
        }
        
        // successor
        if (root == null) {
            return root;
        }
        
        if (root.right != null) {
            root = root.right;
            while (root.left != null) {
                root = root.left;
            }
            return root;
        } else {
            return prev;
        }
    }
}
